### ISSUE TITLE
Enhancement: progress output for `fi defrag`

### ISSUE THREAD

**Issue by Forza-tng** [kdave/btrfs-progs#616]
It would be nice if `btrfs fi defrag` could output progress status when defragmenting large files. The progress could be parsed, and total/remaining extents, for example.

**Comment by kdave** [comment]
There's no way to get the info how the current defragmentation goes. Reading the file mapping periodically could be done but it's not reliable as the defragmented data go through page cache and only after flush the actual extent layout is finalized. Defragmentation is done in cleaner thread, the ioctl only queues the inodes. There can be multiple ioctls started at the same time so there would need to be some handle or context that would provide the info, global defrag state won't do.

**Comment by Forza-tng** [comment]
Oh, I see. Thank you for the explanation. The feature is certainly not that important to warrant changing or creating new interfaces for this. Perhaps we can close this issue?

**Comment by kdave** [comment]
Alternatively you can start the defrag command in steps like

```
btrfs fi defrag -f -s 0 -l 1g file
btrfs fi defrag -f -s 1g -l 1g file
btrfs fi defrag -f -s 2g -l 1g file
...
```

We could add an option that like `btrfs fi defrag --step 1g`. What do you think?

**Comment by Forza-tng** [comment]
That sounds like a good idea. I had forgotten about the start offset option, and having a `--step` option would make it easier to use :) otherwise a shell script wrapper could do the same.

**Comment by Forza-tng** [comment]
@kdave Here's a quick shell script based on your suggestion. I think it works pretty well for my own use-cases.

```shell
#!/bin/sh
#
# defrag_script.sh - A script to defragment large files using btrfs
#

compression_mode="zstd"
comp=""

# Check if running with root privileges
if [ "$(id -u)" -ne 0 ]; then
    echo "This script must be run as root."
    exit 1
fi

# Check if both arguments are provided
if [ $# -lt 2 ] || [ $# -gt 3 ]; then
    echo "Usage: $0 [-c] <blocksize_in_GiB> <filename>"
    exit 1
fi

# Check for the -c flag
if [ "$1" = "-c" ]; then
    comp="-c${compression_mode}"
	echo "Using $compression_mode compression"
    shift
fi

# Check if blocksize input is a valid integer in range 1..4
blocksize="$1"
if [ "$blocksize" -lt 1 ] || [ "$blocksize" -gt 4 ]; then
    echo "Blocksize must be an integer in the range 1..4."
    exit 1
fi

# Check if the file exists
filename="$2"
if [ ! -f "${filename}" ]; then
    echo "File '${filename}' not found."
    exit 1
fi


# Calculate file size in GiB and round up
file_size_in_bytes=$(stat -c %s "${filename}")
file_size_in_GiB=$(( (file_size_in_bytes + blocksize * 1024 * 1024 * 1024 - 1) / (blocksize * 1024 * 1024 * 1024) ))

# Loop through blocks and issue the btrfs command
if [ $file_size_in_GiB -le 1 ]; then
    btrfs filesystem defrag $comp -t ${blocksize}g "${filename}"
else
	trap "echo 'Aborted by user'; exit 1" INT

	# for ((i = 0; i < $file_size_in_GiB; i++)); do
	i=0
	while [ "$i" -lt "$file_size_in_GiB" ]; do
		start_block=$((blocksize * i))

	    echo "Defragmenting block $((i+1)) of $file_size_in_GiB..."
	    btrfs filesystem defrag -f -t ${blocksize}g -czstd -s ${start_block}g -l ${blocksize}g "${filename}"
		errorlevel=$?
		if [ $errorlevel -ne 0 ]; then
			echo "Btrfs reported error code $errorlevel. Aborting..."
			break
		fi
		i=$((i+1))
	done
	trap - INT
fi
echo "Defragmentation complete!"
```

**Comment by kdave** [comment]
`--step` implemented and pushed in devel. I've added that on `-vv` it'll print the steps. Let me know if you'd like to reword the step message.

```
$ btrfs -vv fi defrag --step 1G image
image
defrag range step: start=0 len=1073741824 step=1073741824
defrag range step: start=1073741824 len=1073741824 step=1073741824
defrag range step: start=2147483648 len=1073741824 step=1073741824
defrag range step: start=3221225472 len=1073741824 step=1073741824
```

**Comment by adam900710** [comment]
The `--step` option looks a little overkilled, `-vv` is already enough to me.

**Comment by kdave** [comment]
It's not just reporting the progress, it also flushes the data in each step.

**Comment by Forza-tng** [comment]
I wonder if we should document the use of `-vv` so that users can find this feature (progress output). I can't find it at read the docs or in man pages.

**Comment by kdave** [comment]
Yes this shold be documented.
